chain/network/src/routing.rs

```diff
@@ -752,106 +765,177 @@ pub struct ProcessEdgeResult {
+                let val = if let Some(val) = self.unused.pop() {
+                    self.id2p[val as usize] = peer.clone();
+                    self.used[val as usize] = true;
+                    val
+                } else {
+                    let val = self.id2p.len() as u32;
+                    self.id2p.push(peer.clone());
+                    self.used.push(true);
+                    val
+                };
+                if self.id2p.len() != self.adjacency.len() {
+                    self.adjacency.push(Vec::default());
+                }
+
+                vacant.insert(val);
+                val
+            }
+        }
     }
 
     pub fn add_edge(&mut self, peer0: PeerId, peer1: PeerId) {
+        assert_ne!(peer0, peer1);
         if !self.contains_edge(&peer0, &peer1) {
-            self.add_directed_edge(peer0.clone(), peer1.clone());
-            self.add_directed_edge(peer1, peer0);
+            let id0 = self.get_id(&peer0);
+            let id1 = self.get_id(&peer1);
+
+            self.adjacency[id0 as usize].push(id1);
+            self.adjacency[id1 as usize].push(id0);
+
             self.total_active_edges += 1;
         }
     }
 
     pub fn remove_edge(&mut self, peer0: &PeerId, peer1: &PeerId) {
+        assert_ne!(peer0, peer1);
         if self.contains_edge(&peer0, &peer1) {
-            self.remove_directed_edge(&peer0, &peer1);
-            self.remove_directed_edge(&peer1, &peer0);
+            let id0 = self.get_id(&peer0);
+            let id1 = self.get_id(&peer1);
+
+            self.adjacency[id0 as usize].retain(|&x| x != id1);
+            self.adjacency[id1 as usize].retain(|&x| x != id0);
+
+            self.remove_if_unused(id0);
+            self.remove_if_unused(id1);
+
             self.total_active_edges -= 1;
         }
     }
 
-    // TODO(MarX, #1363): This is too slow right now. (See benchmarks)
     /// Compute for every node `u` on the graph (other than `source`) which are the neighbors of
     /// `sources` which belong to the shortest path from `source` to `u`. Nodes that are
     /// not connected to `source` will not appear in the result.
-    pub fn calculate_distance(&self) -> HashMap<PeerId, HashSet<PeerId>> {
-        let mut queue = vec![];
-        let mut distance = HashMap::new();
-        // TODO(MarX, #1363): Represent routes more efficiently at least while calculating distances
-        let mut routes: HashMap<PeerId, HashSet<PeerId>> = HashMap::new();
-
-        distance.insert(&self.source, 0);
-
-        // Add active connections
-        if let Some(neighbors) = self.adjacency.get(&self.source) {
-            for neighbor in neighbors {
-                queue.push(neighbor);
-                distance.insert(neighbor, 1);
-                routes.insert(neighbor.clone(), vec![neighbor.clone()].drain(..).collect());
-            }
-        }
+    pub fn calculate_distance(&self) -> HashMap<PeerId, Vec<PeerId>> {
+        // TODO add removal of unreachable nodes
```

Comment:
No, it's also not trivial to do this correctly.
Let's say we our node is A.
We add an edge B<->C;
We run calculate_distance that also removes unreachable nodes. Removing B and C.
Then we add A<->B.

This would lead to us to removing a completely valid edges.

Reply:
I added a warning, when number of unreachable nodes is greater than 1000.